### PR TITLE
Jacob

### PR DESCRIPTION
**Pull request by TobiasNorlund**

### CODE REVIEW

File: src/mosse.py [hunk 1]
```diff
@@ -39,7 +39,8 @@ def __init__(self, std=1.0, learning_rate=0.1):
 
     @staticmethod
     def normalize(patch):
-        patch = np.log(1+patch)
+        patch = patch + np.ones(patch.shape)
```
Comment (TobiasNorlund) [inline src/mosse.py:42]: Why was this necessary?

File: src/mosse.py [hunk 2]
```diff
@@ -112,12 +108,27 @@ def detect(self, image):
...
+        patch = crop_patch(image, self.region)
+
+        P = fft2(MOSSETrackerGrayscale.normalize(patch))
+        response = ifft2( np.conjugate(self.M) * P )
+
+        r, c = np.unravel_index(np.argmax(response), response.shape)
+
+        # Keep for visualisation
+        self.last_response = response
+        return self
```
Comment (TobiasNorlund) [inline src/mosse.py:124]: Should return `r, c` here instead?

File: src/run_tracker_interactive.py [hunk 3]
```diff
@@ -9,16 +9,30 @@
...
     parser.add_argument("--seq-idx", type=int, default=0)
     parser.add_argument("--bbox-enlarge-factor", type=float, default=1.0)
+    parser.add_argument("--visualize-filter", type=bool, default=False)
     trackers.add_tracker_args(parser)
     args = parser.parse_args()
 
+    visualize_filter = args.visualize_filter
     tracker = trackers.get_tracker(args)
 
+    if visualize_filter:
```
Comment (TobiasNorlund) [inline src/run_tracker_interactive.py:33]: I visualized the grayscale filters in my notebook here FYI: https://github.com/TobiasNorlund/lfr-module-2/blob/master/notebooks/MOSSE.ipynb

File: src/multi_channel_mosse.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        self.A = np.mean([tracker.A for tracker in self.mosse_trackers], axis=0)
+        self.B = np.mean([tracker.B for tracker in self.mosse_trackers], axis=0)
+
+        self.M = self.A / self.B
+
+        xy = np.mgrid[0:self.region.height, 0:self.region.width].reshape(2,-1).T
+        gaussian = multivariate_normal(mean=np.array(self.region_center), cov=np.eye(2)*self.std**2).pdf(xy)
+        gaussian = np.array(gaussian).reshape(self.region.height, self.region.width)
+        self.last_response = gaussian
```
Comment (TobiasNorlund) [inline src/multi_channel_mosse.py:72]: Not sure I get why we do this? Is it just for having something to visualize?

File: src/multi_channel_mosse.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        d_image = cv2.resize(d_image, (i_h, i_w), interpolation=cv2.INTER_AREA)
+
+        for channel in range(d_channels):
+            self.mosse_trackers.append(MOSSETrackerGrayscale(std=self.std, learning_rate=self.learning_rate))
+            self.mosse_trackers[-1].start(d_image[:,:,channel], region) 
+
+        self.A = np.mean([tracker.A for tracker in self.mosse_trackers], axis=0)
+        self.B = np.mean([tracker.B for tracker in self.mosse_trackers], axis=0)
+
+        self.M = self.A / self.B
```
Comment (TobiasNorlund) [inline src/multi_channel_mosse.py:67]: What is this `self.M` used for?

File: src/multi_channel_mosse.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        d_image = self.descriptor(image)
+        d_image = cv2.resize(d_image, (i_h, i_w), interpolation=cv2.INTER_AREA)
+
+        for i, tracker in enumerate(self.mosse_trackers):
+            tracker.update(d_image[:,:,i])
+
+        self.A = np.mean([tracker.A for tracker in self.mosse_trackers], axis=0)
+        self.B = np.mean([tracker.B for tracker in self.mosse_trackers], axis=0)
+
+        self.M = self.A / self.B
```
Comment (TobiasNorlund) [inline src/multi_channel_mosse.py:115]: I don't see this being used anywhere other than for visualization?

File: .vscode/settings.json [hunk 7]
```diff
@@ -1,4 +1,4 @@
 {
     "terminal.integrated.shell.linux": "/bin/bash",
-    "python.pythonPath": "/usr/local/envs/WASP_rprl_mdl2/bin/python"
+    "python.pythonPath": "/usr/bin/python3"
```
Comment (TobiasNorlund) [inline .vscode/settings.json:3]: Does vscode pick up the right environment with this change?

File: src/multi_channel_mosse.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+def default_descriptor(image):
+    return image
+
+def gradient_descriptor(image):
+    image = np.sum(image, axis=2) / 3
+    laplacian = cv2.Laplacian(image,cv2.CV_64F)[:,:,np.newaxis]
+    sobelx = cv2.Sobel(image,cv2.CV_64F,1,0,ksize=5)[:,:,np.newaxis]
+    sobely = cv2.Sobel(image,cv2.CV_64F,0,1,ksize=5)[:,:,np.newaxis]
+    return np.abs(np.concatenate([sobelx, sobely, laplacian], axis=2))
```
Comment (TobiasNorlund) [inline src/multi_channel_mosse.py:22]: Maybe try adding the grayscale image here as well?

**Review by TobiasNorlund — COMMENTED** [review]
In general I think it looks good. I put some minor comments in the review. @lindbackjacob

File: src/mosse.py (hunk 1 above)
Comment (lindbackjacob) [reply src/mosse.py:42]: I encountered some weird behavior of the normalizing function previously, in which only a subset of dimensions was used when adding the off-set. Consequentially, there was a slight chance of having entries with zeros in the resulting patch, rendering the log output undefined. I used this quick-fix to that problem, but I'm not sure if it still remains. Regardless, they should be equivalent.

File: src/mosse.py (hunk 2 above)
Comment (lindbackjacob) [reply src/mosse.py:124]: Hmm could be. I'll look into this! In that case I would have to make a minor refactoring of the code, but that's not a big deal :)

File: src/multi_channel_mosse.py (hunk 8 above)
Comment (lindbackjacob) [reply src/multi_channel_mosse.py:22]: Could be worth trying! I'll test that
